### DISCUSSION TITLE
Where can I get react version of QAD demo app

### DISCUSSION THREAD

**Discussion by srikanthhub** [clientIO/joint#2082] [Q&A]
### Introduction

Hi all,

I am working on a JointJS task which is similar to the QAD app (https://www.jointjs.com/demos/dialog-generator) which is written in HTML and JS. 
But my requirement is to make it work in a react project. I am facing few issues while converting it to react.
1. Unable to associate an element to its corresponding elementview.
2.  The options ('yes' or 'no') are not loading in Question shape
3. The qad name space is always throwing error

Can anyone help me on getting it work on React application, or share me any project similar to QAD and which is written in React JS, Thanks

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by jamesgeorgewilliams** [comment]
The `dialog generator` currently isn't available in React.

The following demos are available in React:

https://www.jointjs.com/demos/chatbot
https://www.jointjs.com/demos/kitchen-sink-app

https://www.jointjs.com/demos/diagram-index
https://github.com/clientIO/joint-plus-tutorial-react/tree/mui-tree-view

https://www.jointjs.com/demos/tabs
https://github.com/clientIO/joint-plus-tutorial-react/tree/react-tabs

The `qad` namespace doesn't exist in JointJS. The `qad` elements are custom shapes created for this demo.

You can learn more about custom elements, views and namespaces in the following tutorial:
https://resources.jointjs.com/tutorial/ts-shape

In the same manner, you should be able to replicate the setup for the custom `qad` shapes and namespace in React.

**Comment by srikanthhub** [comment]
@jamesgeorgewilliams : Thanks for your reply, I have gone through the links and examples that you have shared. and I am still facing the issue with loading the options in the shape. 

![image](https://user-images.githubusercontent.com/73026917/226615708-8310ca2e-54a2-4b87-92cf-a808772321fe.png)


here is my sample code 
I have created 3 files

1. App.tsx - which creates the graph n paper instance and update the Dom with canvas using React useRef
2. factory.ts - which creates the instance for the shape
3. QadQuestion.tsx - which has shape and its view


### **File 1 : QadQuestion.tsx** 

import * as joint from 'jointjs';
import { V } from "jointjs";
import _ from 'underscore';

export const QadActionType = joint.dia.Element.define('qad.ActionType', {
  optionHeight: 30,
  questionHeight: 45,
  paddingBottom: 30,
  minWidth: 150,
  ports: {
    groups: {
      'in': {
        position: 'top',
        attrs: {
          circle: {
            magnet: 'passive',
            stroke: 'white',
            fill: '#feb663',
            r: 14
          },
          text: {
            pointerEvents: 'none',
            fontSize: 12,
            fill: 'white'
          }
        },
        label: {
          position: {
            name: 'left',
            args: { x: 5 }
          }
        }
      },
      out: {
        position: 'right',
        attrs: {
          'circle': {
            magnet: true,
            stroke: 'none',
            fill: '#31d0c6',
            r: 14
          }
        }
      }
    },
    items: [{
      group: 'in',
      attrs: {
        text: { text: 'in' }
      }
    }]
  },
  attrs: {
    '.': {
      magnet: false
    },
    '.body': {
      width: 'calc(w)',
      height: 'calc(h)',
      rx: '1%',
      ry: '2%',
      stroke: 'none',
      fill: {
        type: 'linearGradient',
        stops: [
          { offset: '0%', color: '#FEB663' },
          { offset: '100%', color: '#31D0C6' }
        ],
        // Top-to-bottom gradient.
        attrs: { x1: '0%', y1: '0%', x2: '0%', y2: '100%' }
      }
    },
    '.btn-add-option': {
      transform: 'translate(10, calc(h - 22))',
      cursor: 'pointer',
      fill: 'white'
    },
    '.btn-remove-option': {
      transform: 'translate(10, 13)',
      cursor: 'pointer',
      fill: 'white'
    },

    // Text styling.
    text: {
      fontFamily: 'Arial'
    },
    '.option-text': {
      fontSize: 11,
      fill: '#000',
      x: 30,
      textVerticalAnchor: 'middle',
    },
    '.question-text': {
      fill: 'white',
      x: 'calc(0.5 * w)',
      y: 15,
      fontSize: 15,
      textAnchor: 'middle',
      textVerticalAnchor: 'top',
      style: {
        textShadow: '1px 1px 0px gray'
      }
    },

    // Options styling.
    '.option-rect': {
      rx: 3,
      ry: 3,
      stroke: 'white',
      strokeWidth: 1,
      strokeOpacity: .5,
      fillOpacity: .5,
      fill: 'white',
      width: 'calc(w)'
    }
  }
}, {

  markup: '<rect class="body"/><text class="question-text"/><g class="options"></g><path class="btn-add-option" d="M5,0 10,0 10,5 15,5 15,10 10,10 10,15 5,15 5,10 0,10 0,5 5,5z"/>',
  optionMarkup: '<g class="option"><rect class="option-rect"/><path class="btn-remove-option" d="M0,0 15,0 15,5 0,5z"/><text class="option-text"/></g>',

  initialize: function(this: any, props: any, smt: any) {
    joint.dia.Element.prototype.initialize.apply(this, props);

    this.on('change:options', this.onChangeOptions, this);
    this.on('change:question', () => {
      this.attr('.question-text/text', this.get('question') || '');
      this.autoresize();
    }, this);

    this.on('change:questionHeight', () => {
      this.attr('.options/transform', `translate(0, ${this.get('questionHeight')})`, { silent: true });
      this.autoresize();
    }, this);

    this.on('change:optionHeight', this.autoresize, this);

    this.attr('.options/transform', `translate(0, ${this.get('questionHeight')})`, { silent: true });
    this.attr('.question-text/text', this.get('question'), { silent: true });

    this.onChangeOptions();
  },

  renderMarkup: function (props: any) {

    // joint.dia.ElementView.prototype.renderMarkup.apply(this, props);

    // A holder for all the options.
    this.$options = this.$('.options');
    // Create an SVG element representing one option. This element will
    // be cloned in order to create more options.
    this.elOption = V(this.model.optionMarkup);

    this.renderOptions();
  },

  onChangeOptions: function() {

    var options = this.get('options');
    var optionHeight = this.get('optionHeight');

    // First clean up the previously set attrs for the old options object.
    // We mark every new attribute object with the `dynamic` flag set to `true`.
    // This is how we recognize previously set attributes.
    var attrs = this.get('attrs');
    _.each(attrs, (attrs, selector) => {

      if (attrs.dynamic) {
        // Remove silently because we're going to update `attrs`
        // later in this method anyway.
        this.removeAttr(selector, { silent: true });
      }
    });

    // Collect new attrs for the new options.
    var offsetY = 0;
    var attrsUpdate = {};
    var questionHeight = this.get('questionHeight');

    console.log(options);
    _.each(options, (option) => {

      var selector = '.option-' + option.id;

      attrsUpdate[selector] = { transform: 'translate(0, ' + offsetY + ')', dynamic: true };
      attrsUpdate[selector + ' .option-rect'] = { height: optionHeight, dynamic: true };
      attrsUpdate[selector + ' .option-text'] = { text: option.text, dynamic: true, transform: `translate(0, ${optionHeight / 2})` };

      offsetY += optionHeight;

      var portY = offsetY - optionHeight / 2 + questionHeight;
      if (!this.getPort(option.id)) {
        this.addPort({ group: 'out', id: option.id, args: { y: portY }});
      } else {
        this.portProp(option.id, 'args/y', portY);
      }
    });

    this.attr(attrsUpdate);
    this.autoresize();
  },

  autoresize: function() {

    var options = this.get('options') || [];
    var gap = this.get('paddingBottom') || 20;
    var height = options.length * this.get('optionHeight') + this.get('questionHeight') + gap;
    var width = measureText(this.get('question'), {
      fontSize: this.attr('.question-text/fontSize')
    }).width;
    this.resize(Math.max(this.get('minWidth') || 150, width), height);
  },

  addOption: function(option) {

    var options = JSON.parse(JSON.stringify(this.get('options')));
    options.push(option);
    this.set('options', options);
  },

  removeOption: function(id) {

    var options = JSON.parse(JSON.stringify(this.get('options')));
    this.removePort(id);
    this.set('options', _.without(options, _.find(options, { id: id })));
  },

  changeOption: function(id, option) {

    if (!option.id) {
      option.id = id;
    }

    var options = JSON.parse(JSON.stringify(this.get('options')));
    options[_.findIndex(options, { id: id })] = option;
    this.set('options', options);
  }
});

export const QadActionTypeView = joint.dia.ElementView.extend({
model: QadActionType,
  events: {
    'click .btn-add-option': 'onAddOption',
    'click .btn-remove-option': 'onRemoveOption'
  },

  presentationAttributes: joint.dia.ElementView.addPresentationAttributes({
    options: ['OPTIONS']
  }),

  confirmUpdate: function(flags, props: any) {

    joint.dia.ElementView.prototype.confirmUpdate.apply(this, props);
    if (this.hasFlag(flags, 'OPTIONS')) {
      this.renderOptions();
    }
  },

  renderMarkup: function(this: any, props: any) {

    // joint.dia.ElementView.prototype.renderMarkup.apply(this, props);

    // A holder for all the options.
    this.$options = document.querySelector('.options');
    // Create an SVG element representing one option. This element will
    // be cloned in order to create more options.
    this.elOption = V(this.model.optionMarkup);

    this.renderOptions();
  },

  renderOptions: function() {

    this.$options.empty();

    _.each(this.model.get('options'), (option, index) => {

      var className = 'option-' + option.id;
      var elOption = this.elOption.clone().addClass(className);
      elOption.attr('option-id', option.id);
      this.$options.append(elOption.node);

    });

    // Apply `attrs` to the newly created SVG elements.
    this.update();
  },

  onAddOption: function() {

    this.model.addOption({
      id: _.uniqueId('option-'),
      text: 'Option ' + this.model.get('options').length
    });
  },

  onRemoveOption: function(evt) {

    this.model.removeOption(V(evt.target.parentNode).attr('option-id'));
  }
});

function measureText (text, attrs) {

  var fontSize = parseInt(attrs.fontSize, 10) || 10;

  var svgDocument = V('svg').node;
  var textElement = V('<text><tspan></tspan></text>').node;
  var textSpan = textElement.firstChild;
  var textNode = document.createTextNode('');

  textSpan?.appendChild(textNode);
  svgDocument.appendChild(textElement);
  document.body.appendChild(svgDocument);

  var lines = text.split('\n');
  var width = 0;

  // Find the longest line width.
  _.each(lines, function(line) {

    textNode.data = line;
    var lineWidth = 100;

    width = Math.max(width, lineWidth);
  });

  var height = lines.length * (fontSize * 1.2);

  V(svgDocument).remove();

  return { width: width, height: height };
};


### **File 2 : factory.ts** 

import { QadActionType } from './QadObservation';

export const createActiveType = (model) => {
    var title = model.contentTitle;
    var trimText = title.length > 25 ? title.substring(0, 25) + " ..." : title;
    var q = new QadActionType({
      position: { x: 400 - 50, y: 30 },
      size: { width: 100, height: 70 },
      question: trimText,
      contentType: "test",
      inPorts: [{ id: "in", label: "In" }],
      options: [
        { "id": "Normal", "text": "Normal" },
        { "id": "Low", "text": "Low" },
        { "id": "Medium", "text": "Medium" },
        { "id": "High", "text": "High" },
      ]
    });
  
    return q;
};

### File 3: App.tsx 

const App: FC<IProps> = () => {
    let graph: any = new dia.Graph({}, { cellNamespace: shapes });
    const canvas: any = useRef(null);
    // Define paper model
    let paper: any = new dia.Paper({
        model: graph,
        cellViewNamespace: shapes,
        width: 1000,
        height: 1000,
        gridSize: 10,
        linkPinning: false,
        drawGrid: true,
        defaultLink: new dia.Link({
            router: { name: 'manhattan' },
            connector: { name: 'rounded' },
            attrs: {
                '.marker-target': { d: 'M 10 0 L 0 5 L 10 10 z', fill: '#6a6c8a', stroke: '#6a6c8a' },
                '.connection': {
                    stroke: '#6a6c8a', 'stroke-width': 2
                }
            }
        }),
        async: true,
        sorting: dia.Paper.sorting.APPROX,
        interactive: { linkMove: false },
        snapLinks: { radius: 75 },
        defaultConnectionPoint: { name: 'boundary' }
    });;

    useEffect(() => {
        // Define graph
        // graph = new dia.Graph({}, { cellNamespace: shapes });

        // Define paper scroller model
        const paperScroller = new ui.PaperScroller({
            paper:paper,
            cursor: 'grab'
        });

        // Append and re-center paperScoller to canvas
         canvas.current.appendChild(scroller.el);

        paperScroller.render();

        //paper.findViewByModel(QadSingleSelectQuestion).render();
        // Action pointerdown on paper


        // Action pointerclick for elements on paper
        paper.on('element:pointerclick', (elementView) => {
            const halo = new ui.Halo({
                cellView: elementView,
                type: 'toolbar',
                theme: 'modern',
                boxContent: false,
                // handles
            }).render();

            halo.removeHandle('resize');
            halo.removeHandle('link');
            halo.removeHandle('fork');
            halo.removeHandle('rotate');
        });

        // Action pointerclick for links on paper
        paper.on('link:pointerclick', (linkView) => {
            paper.removeTools();
            const toolsView = new dia.ToolsView({
                tools: [
                    new linkTools.Vertices(),
                    new linkTools.SourceArrowhead(),
                    new linkTools.TargetArrowhead(),
                    new linkTools.Segments,
                    new linkTools.Remove({ offset: -20, distance: 40 })
                ]
            });
            linkView.addTools(toolsView);
        });
     


        // React on changes in the graph.
        graph.on('change add remove', () => {
            const diagramJSONString = JSON.stringify(graph.toJSON());
         //   console.log('Diagram JSON', diagramJSONString);
        });

        // Remove all links
        const links = graph.getLinks();
        links.forEach((link) => link.remove());     

            createActiveType({
                contentTitle: "Take observation",
                contentType: "Action"
            }).addTo(graph)

            
            graph.addCell(cell);

        });

       paper.render();

    }, [questionTypes]);


    return (
        <div className="canvas" ref={canvas} />
    );
}

export default App;

**Comment by jamesgeorgewilliams** [comment]
It is a little difficult to follow the code here as it doesn't follow these guidelines:
https://github.com/clientIO/joint/wiki/Asking-Technical-Questions

Could you send a functional example to `support@jointjs.com`, and we will be able to take a look. You can send all project files as a ZIP file. Thanks

**Comment by srikanthhub** [comment]
I have emailed the project files, please take a look. Thanks

**Comment by srikanthhub** [comment]
Thanks @jamesgeorgewilliams for your quick help. Now we can able apply the same changes in my project, and it is working as expected. Thank you :)
